Socket security updates

NthTensor opened the pull request:
Fixes problems 1 and 2 of #810. Also contains a script for browser debugging.

cozuya commented on routes/socket/game/election-util.js:
```diff
@@ -12,10 +12,10 @@ module.exports.selectChancellor = (socket, data) => {
 		!game.private.seatedPlayers ||
-		!socket.passport.handshake.session ||
-		!game.publicPlayersState.find(player => player.userName === socket.passport.handshake.session).length ||
+		!socket.handshake.session.passport ||
```
woops

cozuya commented on routes/socket/routes.js:
```diff
@@ -53,6 +53,27 @@ const gamesGarbageCollector = () => {
+	return false;
+};
+
+const ensureIngame = (passport, uid) => {
```
yeah this is a good idea

cozuya commented:
sounds good thanks.  Can you try to resolve conflicts?  If not I can take a look tonight.  I'm looking to release this today.

NthTensor commented:
Yep. Its about half way debugged RN. I will @ you in discord when its time.

NthTensor commented:
# Status Update
 All routes should now be secure. However, this branch is still unstable.
## Overview
+ Removed the `data.userName` pattern from all messages.
+ Changed `routes.js` to help manage authentication (`ensureAuthenticated`, `ensureInGame`). 
+ Moved additions to `checkUserStatus`. *Though if I am honest, this has not worked out too well, and probably needs a second look.*
+ Did some other housekeeping and bug fixing.
+ Added some `AddNewGame` guards. -**Hexicube** 
+ Updated `DevGame` to use the new packet format.
+ Added a guard to president and chancellor actions (see below).
```
const handle = (passport, game, data) => {
    const { presidentIndex } = game.gameState;
    const president = seatedPlayers[presidentIndex];
    if (president.userName !== passport.user) {
        return;
    }
}
``` 
## Known Bugs
- [ ] GeneralChat doesn’t work
- [ ] GameChat doesn’t work for observers
- [ ] GameList acting strangely
- [ ] SelectSpecialElection is not secure, but it will take some work to fix it.

NthTensor commented:
I will watch out for that. When I was looking at the code, I noticed a lot of redundancy in the defensive statements. I thought it might be easier maintain if the redundancy was removed, and I implemented the authorization helpers in `routes.js` to help with that. That this was probably beyond the scope of this PR, and I probably need to add more explicit guards to the `routes.js` defensive helper functions. If this proves to be to unstable in testing, I will revert the handlers, so as to have defensive code in both them and `routes.js`.

Also, I think issue in #81 where users do not appear in UserList should be resolved by this PR, although that part is still a bit buggy.

cozuya commented:
Sounds good looking forward to taking a look.

cozuya commented:
@NthTensor is this review for review and merge to master?

cozuya commented:
oops didnt see your checklist

cozuya commented:
also tournys wont work and I've disabled them.  Will take more time to make those work but would like to get all the crash bugs released to prod first.

NthTensor commented:
I made that checklist thinking I would be able to edit it. Apparently you can only edit comments on repos you belong to. -_-

There are is only one remaining known bug, and its not a crash. Remake brings players to the lobby, instead of a new game. Our changes to the leaveGame handler appear to be the problem, but I do not really understand how you guys set up remakes. Will need your help for that one.

Also, SelectSpecialElection is still not secure, but that will take a fairly large redesign.

Other than that this is O.K. for review.

cozuya commented on routes/socket/routes.js:
```diff
@@ -147,17 +204,21 @@ module.exports = () => {
+				}
 			})
-			.on('getModInfo', count => {
+			.on('getModInfo', count => { // I cant tell if this needs to be secure or not
```
these don't need security, if so I have bigger problems.

cozuya commented:
I'm seeing that observers can't chat at all and I have 1 user who can chat in general chat and one who can't 😑

NthTensor commented:
I neglected to test observer functionality. Pushing fix asap.

cozuya commented:
awesome.. the genchat thing is cause one user had no games on laptop 🤦‍♂️

cozuya commented:
I've merged this manually to master and will deploy shortly tonight.  Thanks for all your help!  If we missed something no big deal can catch that on a case by case basis.
